kedro/extras/datasets/pandas/sql_dataset.py

```diff
@@ -207,42 +207,53 @@ def __init__(
         self._load_args["table_name"] = table_name
         self._save_args["name"] = table_name
 
-        self._load_args["con"] = self._save_args["con"] = credentials["con"]
+        self._connection_str = credentials["con"]
+        self.create_connection(self._connection_str)
+
+    @classmethod
+    def create_connection(cls, connection_str: str) -> None:
+        """Given a connection string, create singleton connection
+        to be used across all instances of `SQLTableDataSet` that
+        need to connect to the same source.
+        """
+        if connection_str in getattr(cls, "engines", {}):
+            return
+
+        engines = cls.engines if hasattr(cls, "engines") else {}  # type:ignore
```

Comment:
This feels a bit weird to me 🤔  Probably I'm missing something here... but why not just set `engines = {}` with other class attributes above `def __init__`? It means setting a mutable default argument but I think that's fine here (and we already do it with the other class attributes there, although that is arguably a bit different).

Then we could just simplify like:
```
if connection_str in cls.engines:
    return
...
cls[engines] = engine # no need for local `engines` object at all
```

Reply:
I don't follow your suggestion sorry. I've never seen syntax like `cls[engines]` before either, it doesn't look like idiomatic python to me.

Reply:
Oh sorry, that was a typo. I mean `cls.engines`. What I meant was this:
```
class SQLTableDataSet(AbstractDataSet):
    engines = {} # like we do with DEFAULT_LOAD/SAVE_ARGS

    def __init__():
        # as you have it now

    @classmethod
    def create_connection(cls, connection_str: str) -> None:
        if connection_str in cls.engines: 
            return

        try:
            engine = create_engine(connection_str)
        except ImportError as import_error:
            raise _get_missing_module_error(import_error) from import_error
        except NoSuchModuleError as exc:
            raise _get_sql_alchemy_missing_error() from exc

        cls.engines[connection_str] = engine
```

So no need for the `hasattr`/`getattr` stuff.

Reply:
Makes sense, I was probably just overthinking. Done!